src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -4035,7 +4035,7 @@ trait Types
     }
 
     def check(tp1: Type, tp2: Type) = (
-      if (tp1.typeSymbol.isClass && tp1.typeSymbol.hasFlag(FINAL))
+      if (tp1.typeSymbol.isClass && tp1.typeSymbol.hasFlag(FINAL) && tp1 != SingletonClass.tpe)
```

Comment:
A few similar checks (e.g `CheckabilityChecker` are predicated on `isEffectivelyFinal`. I think we should also change that, too.

I'd also recommend here:

```
val sym1 = tp1.typeSymbol
if (sym1.isClass && sym1.hasFlag(FINAL) && sym1 != SingletonClass)
```

Reply:
Consider it done ...